File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    int expand(string &s, int left, int right) {

            while (left >= 0 && right < s.size() && s[left] == s[right]) {
                left--;
                right++;
            }

            return right - left - 1;
    }
    
    string longestPalindrome(string s) {
        /*int n = s.size();

        int start = 0;
        int maxLen = 1;

        for (int i = 0; i < n; i++) {

            // Odd length palindrome
            int left = i;
            int right = i;

            while (left >= 0 && right < n && s[left] == s[right]) {
                if (right - left + 1 > maxLen) {
                    start = left;
                    maxLen = right - left + 1;
                }

                left--;
                right++;
            }

            // Even length palindrome
            left = i;
            right = i + 1;

            while (left >= 0 && right < n && s[left] == s[right]) {
                if (right - left + 1 > maxLen) {
                    start = left;
                    maxLen = right - left + 1;
                }

                left--;
                right++;
            }
        }

        return s.substr(start, maxLen);*/
        int start = 0;
        int maxLen = 0;

        for (int i = 0; i < s.size(); i++) {

            // Odd length palindrome
            int len1 = expand(s, i, i);

            // Even length palindrome
            int len2 = expand(s, i, i + 1);

            int len = max(len1, len2);

            if (len > maxLen) {
                maxLen = len;

                // Find starting index
                start = i - (len - 1) / 2;
            }
        }

        return s.substr(start, maxLen);
        
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    private int expand(String s, int l,int r){
        while(l>=0 && r<s.length() && s.charAt(l)==s.charAt(r)){
            l--;
            r++;
        }
        return r-l-1;
    }
    public String longestPalindrome(String s) {
        int end= 0;
        int start = 0;
        for(int i=0;i<s.length();i++){
            int odd = expand(s,i,i);
            int even = expand(s,i,i+1);
            int len = Math.max(odd,even);
            if(len>end-start){
                end=i+len/2;
                start = i-(len-1)/2;
            }
        }
        return s.substring(start,end + 1);
    }
}